### PR TITLE
feat: add amino signing support

### PR DESCRIPTION
**Pull request by ryanchristo**
Closes: #10

This pull request adds support for amino signing.

- [x] `MsgCreateGroupPolicy`
- [x] `MsgCreateGroupWithPolicy` (`policyAsAdmin` set to `true`)
- [x] `MsgCreateGroupWithPolicy` (`policyAsAdmin` set to `false`)
- [x] `MsgExec`
- [x] `MsgSubmitProposal` (no messages)
- [x] `MsgSubmitProposal` (with `MsgSend`)
- [x] `MsgSubmitProposal` (with `MsgDelegate`)
- [x] `MsgSubmitProposal` (with `MsgRedelegate`)
- [x] `MsgSubmitProposal` (with `MsgUndelegate`)
- [x] `MsgSubmitProposal` (with `MsgWithdrawDelegatorReward`)
- [x] `MsgSubmitProposal` (with `MsgUpdateGroupMetadata`)
- [x] `MsgSubmitProposal` (with `MsgUpdateGroupMembers`)
- [x] `MsgSubmitProposal` (with `MsgUpdateDecisionPolicy`)
- [x] `MsgUpdateGroupMetadata`
- [x] `MsgUpdateGroupMembers`
- [x] `MsgUpdateDecisionPolicy`
- [x] `MsgVote` (`exec` set to `1`)
- [x] `MsgVote` (`exec` set to `0`)

Deploy Preview: https://deploy-preview-69--regen-groups-ui.netlify.app/

### CODE REVIEW

**Comment by ryanchristo** [comment]
Note to self. The current setup using `getOfflineSignerOnlyAmino` works with messages that do not have embedded types but it does not work with creating and update group policies and submitting proposals. Still investigating.

**Comment by ryanchristo** [comment]
`MsgCreateGroupWithPolicy` (using new helpers)

```
Group could not be created:
Broadcasting transaction failed with code 4 (codespace: sdk). Log: signature verification failed; please verify account number (1), sequence (13) and chain-id (regenlocal): unauthorized
```

`MsgSubmitProposal` (not using new helpers)

```
Broadcasting transaction failed with code 7 (codespace: sdk). Log: msg 0: invalid from address: empty address string is not allowed: invalid address
```

Updating group name and metadata works as expected.

**Comment by ryanchristo** [comment]
`MsgSubmitProposal` (not using any helpers and no message)

```
Broadcasting transaction failed with code 4 (codespace: sdk). Log: signature verification failed; please verify account number (1), sequence (3) and chain-id (regenlocal): unauthorized
```

Same error as `MsgCreateGroupWithPolicy` with helpers.

File: package.json [hunk 1]
```diff
@@ -20,16 +20,16 @@
...
+    "@cosmjs/stargate": "0.29.5",
+    "@cosmjs/tendermint-rpc": "^0.29.5",
     "@emotion/react": "^11.10.8",
     "@emotion/styled": "^11.10.8",
     "@hookform/resolvers": "^3.1.0",
     "@keplr-wallet/cosmos": "^0.11.59",
     "@keplr-wallet/stores": "^0.11.59",
     "@osmonauts/helpers": "^0.7.0",
-    "@regen-network/api": "^1.0.0-alpha1",
+    "@regen-network/api": "../regen-js/packages/api",
```
Comment (ryanchristo) [inline package.json:32]: Using https://github.com/regen-network/regen-js/pull/75 to test

File: src/api/group.messages.ts [hunk 2]
```diff
@@ -20,9 +21,11 @@ export function msgCreateGroupWithPolicy(values: GroupWithPolicyFormValues) {
     policyType,
     votingWindow,
   } = values
-  return GroupMsgWithTypeUrl.createGroupWithPolicy({
+  // NOTE: We use the encoded msg type to support amino signing with nested types.
+  // See https://github.com/osmosis-labs/telescope/issues/281
+  const encodedMsg: MsgCreateGroupWithPolicyEncoded = {
     admin,
-    groupPolicyMetadata: '',
+    groupPolicyMetadata: ' ', // NOTE: whitespace required for amino
```
Comment (ryanchristo) [inline src/api/group.messages.ts:28]: Adding this whitespace resolved the issue with this message.

File: src/api/group.messages.ts [hunk 3]
```diff
@@ -40,10 +43,15 @@ export function msgCreateGroupWithPolicy(values: GroupWithPolicyFormValues) {
     members: members.map((m) => ({
       address: m.address,
       weight: m.weight.toString(),
-      metadata: '',
+      metadata: ' ', // NOTE: whitespace required for amino
```
Comment (ryanchristo) [inline src/api/group.messages.ts:46]: Adding this whitespace resolved the issue with this message.

**Comment by ryanchristo** [comment]
`MsgCreateGroupWithPolicy` is working after updating metadata fields to not be empty and only if `groupPolicyAsAdmin` is set to `true`. `MsgSubmitProposalEncoded` type does not exist so trying to figure out why and how to proceed.

File: src/api/group.messages.ts [hunk 4]
```diff
@@ -20,10 +21,12 @@ export function msgCreateGroupWithPolicy(values: GroupWithPolicyFormValues) {
...
     votingWindow,
   } = values
-  return GroupMsgWithTypeUrl.createGroupWithPolicy({
+  // NOTE: We use the encoded msg type to support amino signing with nested types.
+  // See https://github.com/osmosis-labs/telescope/issues/281
+  const encodedMsg: MsgCreateGroupWithPolicyEncoded = {
     admin,
-    groupPolicyMetadata: '',
-    groupPolicyAsAdmin: policyAsAdmin === 'true',
+    groupPolicyMetadata: ' ', // TODO: unauthorized error when empty
```
Comment (blushi) [inline src/api/group.messages.ts:28]: have you tried with undefined instead (using `clearEmptyStr`)?
I remember we were doing something similar in the amino converters within the ts-proto based regen-js code: https://github.com/regen-network/regen-js/blob/022215439e0db0771e5968d4532fa68630096521/packages/api/src/tx/modules/ecocredit/converter-utils.ts#L7

**Comment by ryanchristo** [comment]
This is currently working with all group messages but the project won't build due to typescript errors.

**Comment by ryanchristo** [comment]
I resolved the ts errors within this project and specifically the ts errors in regen-js that were showing up here:

https://github.com/regen-network/regen-js/pull/79

With this pull request and a new alpha tag, we should have amino support fully operational.

**Comment by ryanchristo** [comment]
I think we should manually fix the typescript errors upstream (https://github.com/regen-network/regen-js/pull/79) and tag a new version of the api package that we can use until we resolve the issues in either regen-js or telescope. I think it's ok if the changes get overwritten as we can always pull them back in if we need to tag a new release of the api package that will be used in the groups ui. I'd like to at least have a version we know can use for the MVP.

**Comment by ryanchristo** [comment]
If we decide to merge https://github.com/regen-network/regen-js/pull/79, I'll tag another release and update and then open for review.

**Comment by ryanchristo** [comment]
This is now ready to review.

**Comment by ryanchristo** [comment]
And follow up issue create #105.

**Comment by blushi** [comment]
I'm getting an error when editing group:
`Broadcasting transaction failed with code 4 (codespace: sdk). Log: signature verification failed; please verify account number (12), sequence (1866) and chain-id (regen-redwood-1): unauthorized`

**Comment by ryanchristo** [comment]
Oof. Looking into it. Also noticed the update group messages need to be added. Will fix and add.

**Comment by ryanchristo** [comment]
The new messages require a new file from the api `src` directory so we may need to manually resolve and release a version with the manual fixes upstream...

**Comment by ryanchristo** [comment]
Also running into issues now with submitting proposals with the following messages:

- MsgUpdateGroupMembers
- MsgUpgradeGroupPolicyDecisionPolicy

**Comment by blushi** [comment]
Working now with a `MsgUpdateGroupMembers` proposal.

But I still get the following error when trying to create a proposal with `MsgUpdateGroupDecisionPolicy` as soon as I have a percentage decision policy:

> Proposal could not be created:
> Broadcasting transaction failed with code 10 (codespace: math). Log: msg 0: decision policy: threshold: parse mantissa: : invalid decimal string: invalid decimal string

Here's a fix: https://github.com/regen-network/groups-ui/pull/116

**Review by blushi — APPROVED** [review]
pre-approving, good to merge after final testing

**Comment by ryanchristo** [comment]
> pre-approving, good to merge after final testing

Ran through submit proposal messages one more time with each supported message (and no messages) on redwood:

https://deploy-preview-69--regen-groups-ui.netlify.app/37

**Pull request merged**